bot/exts/filtering/_filters/antispam/role_mentions.py

```diff
@@ -0,0 +1,42 @@
+from datetime import timedelta
+from itertools import takewhile
+from typing import ClassVar
+
+import arrow
+from pydantic import BaseModel
+
+from bot.exts.filtering._filter_context import Event, FilterContext
+from bot.exts.filtering._filters.filter import UniqueFilter
+
+
+class ExtraRoleMentionsSettings(BaseModel):
+    """Extra settings for when to trigger the antispam rule."""
+
+    interval_description: ClassVar[str] = (
+        "Look for rule violations in messages from the last `interval` number of seconds."
+    )
+    threshold_description: ClassVar[str] = "Maximum number of role mentions before the filter is triggered."
+
+    interval: int = 10
+    threshold: int = 3
+
+
+class DuplicatesFilter(UniqueFilter):
+    """Detects too many role mentions sent by a single user."""
+
+    name = "role_mentions"
+    events = (Event.MESSAGE,)
+    extra_fields_type = ExtraRoleMentionsSettings
+
+    async def triggered_on(self, ctx: FilterContext) -> bool:
+        """Search for the filter's content within a given context."""
+        earliest_relevant_at = arrow.utcnow() - timedelta(seconds=self.extra_fields.interval)
+        relevant_messages = list(takewhile(lambda msg: msg.created_at > earliest_relevant_at, ctx.content))
+        detected_messages = {msg for msg in relevant_messages if msg.author == ctx.author}
+        total_recent_mentions = sum(len(msg.role_mentions) for msg in relevant_messages)
```

Comment:
This stuff seems kind of repetitive (I think there was at least 4 or 5 usages across the filters), perhaps we can provide some superclass instance function to do this or some default implementation?

Reply:
This will have to assume that every antispam filter has an extra setting called `interval` (holds now, but isn't logically guaranteed), and the attribute being summed is different every time, so you either exclude the fourth line, or send a function to fetch the right value from each message. I'll think about it, but at first glance trying to generalize it seems clumsy.